README.md
<!-- markdownlint-disable MD033 MD003 -->

Scorch<a id="footnote-0-1-backref" href="#footnote-0-1">¹</a>
======

[![Build Status](https://github.com/LoicGrobol/scorch/workflows/CI/badge.svg)](https://github.com/LoicGrobol/scorch/actions?query=workflow%3ACI)
[![PyPI](https://img.shields.io/pypi/v/scorch.svg)](https://pypi.org/project/scorch)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

This is an alternative implementation of the coreference scorer for the CoNLL-2011/2012 shared tasks
on coreference resolution.

It aims to be more straightforward than the [reference implementation][ref-scorer], while
maintaining as much compatibility with it as possible.

The implementations of the various scores are as close as possible from the formulas used by
<a href="#pradhan2014scoring">Pradhan et al. (2014)</a>, with the edge cases for BLANC taken from
<a href="recasens2011BLANC">Recasens and Hovy (2011)</a>.

---
<sub><a id="footnote-0-1" href="#footnote-0-1-backref">1.</a> **S**corer for **cor**eference **ch**ains.</sub>

[ref-scorer]: https://github.com/conll/reference-coreference-scorers

## Usage

```bash
scorch gold.json sys.json out.txt
```

## Install

From the cheeseshop

```bash
python3 -m pip install --user scorch
```

Or directly from git

```bash
python3 -m pip install --user git+https://github.com/LoicGrobol/scorch.git
```

## Formats

### Single document

The input files should be JSON files with a `"type"` key at top-level

- If `"type"` is `"graph"`, then top-level should have at top-level
  - A `"mentions"` key containing a list of all mention identifiers
  - A `"links"` key containing a list of pairs of corefering mention identifiers
- If `"type"` is `"clusters"`, then top-level should have a `"clusters"` key containing a mapping
  from clusters ids to cluster contents (as lists of mention identifiers).

Of course the system and gold files should use the same set of mention identifiers for the mentions
they have in common.

For convenience, the [`conll.py`](/scorch/conll.py) converts CoNLL-2012 files to this format.

### Multiple documents

If the inputs are directories, files with the same base name (excluding extension) as those present
in the sys directory are expected to be present in the gold directory, with exactly one gold file
for each sys file.
In that case, the output scores will be the micro-average of the individual files
scores, ie their arithmetic means weighted by the relative numbers of

- Gold mentions for Recall
- System mentions for Precision
- The sum of the previous two for F₁

This is different from the reference interpretation where

- **MUC** weighting ignores mentions in singleton entities
  - This should not make any difference for the CoNLL-2012 dataset, since singleton entities are not
    annotated.
  - For datasets with singletons, the shortcomings of MUC are well known, so this score shouldn't
    matter much
- **BLANC** is calculated by micro-averaging coreference and non-coreference separately, using
  the number of links as weights instead of the number of mentions.
  - This is roughly equivalent to weighting coreference scores per document by their number of
    non-singleton clusters and non-coreference scores by the square of their number of mentions.
    This give disproportionate importance to large documents, which is not desirable
    in heterogenous corpora

The CoNLL average score is the arithmetic mean of the global MUC, B³ and CEAFₑ F₁ scores.

## Sources

- <a id="pradhan2014scoring" />**Scoring Coreference Partitions of Predicted Mentions: A Reference
  Implementation.** Sameer Pradhan, Xiaoqiang Luo, Marta Recasens, Eduard Hovy, Vincent Ng and
  Michael Strube. *Proceedings of the 52nd Annual Meeting of the Association for Computational
  Linguistics*, Baltimore, MD, June 2014. ([pdf](http://aclweb.org/anthology/P/P14/P14-2006.pdf))
- <a id="recasens2011BLANC" />**BLANC: Implementing the Rand Index for Coreference Evaluation.**
  Marta Recasens and Eduard Hovy In: *Natural Language Engineering* 17 (4). Cambridge University
  Press, 2011.
  ([pdf](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.300.9229&rep=rep1&type=pdf))
- <a id="luo2014BLANC" /> **An Extension of BLANC to System Mentions.** Xiaoqiang Luo, Sameer
  Pradhan, Marta Recasens and Eduard Hovy. *Proceedings of the 52nd Annual Meeting of the
  Association for Computational Linguistics*, Baltimore, MD, June 2014.
  ([pdf](http://aclweb.org/anthology/P/P14/P14-2005.pdf))
- The version of the Kuhn-Munkres algorithm used for the CEAF scores uses
  [`scipy.optimize.linear_sum_assignment`](https://docs.scipy.org/doc/latest/reference/generated/scipy.optimize.linear_sum_assignment.html),
  with `$-ϕ_n$` as cost function.

## License

Unless otherwise specified (see <a href="#license-exceptions">below</a>), the following licence (the
so-called “MIT License”) applies to all the files in this repository.
See also [LICENSE.md](LICENSE.md).

```text
Copyright 2018 Loïc Grobol <loic.grobol@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
associated documentation files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

### <a id="license-exceptions">License exceptions</a>

- The reference scorer testcases located in
  [`tests/fixtures/conll/`](tests/fixtures/conll/datafiles) are distributed under the [Creative
  Commons Attribution ShareAlike 4.0 International
  License](http://creativecommons.org/licenses/by-sa/4.0/)
  - **Copyright** © 2011- Sameer Pradhan pradhan \<at\> cemantix.org
  - **Authors**
    - Emili Sapena, Universitat Politècnica de Catalunya, <http://www.lsi.upc.edu/~esapena>, esapena
      \<at\> lsi.upc.edu
    - Sameer Pradhan, <http://cemantix.org>, pradhan \<at\> cemantix.org
    - Sebastian Martschat, sebastian.martschat \<at\> h-its.org
    - Xiaoqiang Luo, xql \<at\> google.com
  - **Origin** <http://conll.github.io/reference-coreference-scorers>
  
  These files are taken verbatim from the 8.0.1 of the official CoNLL scorer at <https://github.com/conll/reference-coreference-scorers/releases>


CHANGELOG.md
Changelog
=========

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] (0.1.0)

[Unreleased]: https://github.com/loicgrobol/scorch/compare/v0.2.0...HEAD

## [0.2.0] — 2020-10-28

[0.2.0]: https://github.com/loicgrobol/scorch/compare/v0.1.0...v0.2.0

### Changed

- Slightly faster BLANC ([#13](https://github.com/LoicGrobol/scorch/pull/13))

## [0.1.0] — 2020-10-12

[0.1.0]: https://github.com/loicgrobol/scorch/compare/v0.0.26...v0.1.0

### Fixed

- Stop redirecting SIGPIPE
  [incorrectly](https://docs.python.org/3/library/signal.html#note-on-sigpipe)
- B³ and CEAF P and R are now set to 0 if their denominator is 0, following the reference implementation ([@Rexhaif](https://github.com/Rexhaif) in [#10](https://github.com/LoicGrobol/scorch/pull/10))

## [0.0.26] - 2019-12-16

[0.0.26]: https://github.com/loicgrobol/scorch/compare/v0.0.25...v0.0.26

### Added

- [`speedtest.py`](/tests/speedtest.py), a benchmark tool for the speed of our implementations

### Changed

- Speed up of BLANC computation by a factor of 75 at the price of straightforwardness of
  implementation. The slow and straightforward implementation is still available behind a parameter
  and is used to test the consistency of the fast one.
  ([#5](https://github.com/LoicGrobol/scorch/pull/5))

## Previous versions

No real changelog was kept for the previous versions, you can have a look at the history but it is
not for the faint of heart


tests/utils.py
import hypothesis
from hypothesis import strategies as st


@st.composite
def clusterings(draw, max_size=None, allow_singletons=True):
    n_elements = draw(st.integers(min_value=1, max_value=max_size))
    n_clusters = draw(st.integers(min_value=1, max_value=n_elements))
    affectations = draw(
        st.lists(
            st.integers(min_value=0, max_value=n_clusters - 1),
            min_size=n_elements,
            max_size=n_elements,
        )
    )
    clusters = [[] for _ in range(n_clusters)]
    for i, n in enumerate(affectations):
        clusters[n].append(i)
    # FIXME: this removes empty clusters but it would be better to avoid generating them at all
    res = [set(c) for c in clusters if c]
    if not allow_singletons:
        res = [c for c in clusters if len(c) > 1]
    hypothesis.assume(res)
    return res


tests/speedtest.py
import pathlib
import timeit
import typing as ty

from scorch import scores
from scorch import main as scorch


tests_dir = pathlib.Path(__file__).resolve().parent

Clustering = ty.Sequence[ty.Set[ty.Hashable]]
Metric = ty.Callable[[Clustering, Clustering], ty.Tuple[float, float, float]]


def test_metrics(
    clusters: Clustering,
    num_runs: int = 100,
    repeat: int = 5,
    metrics: ty.Dict[str, Metric] = scorch.METRICS,
):
    print(f"Testing metrics speed: {num_runs} calls, best of {repeat}")
    print("metric\ttotal\tper call")
    for (name, fun) in metrics.items():
        runtime = min(
            timeit.repeat(
                "fun(clusters, clusters)",
                globals={"fun": fun, "clusters": clusters},
                number=num_runs,
                repeat=repeat,
            )
        )
        print(f"{name}\t{runtime}\t{runtime/num_runs}")


def test_blanc_speedup(clusters: Clustering, num_runs: int = 100, repeat: int = 5):
    print(f"Testing BLANC speedup: {num_runs} calls, best of {repeat}")
    slow_runtime = min(
        timeit.repeat(
            "blanc(clusters, clusters, False)",
            globals={"blanc": scores.blanc, "clusters": clusters},
            number=num_runs,
            repeat=repeat,
        )
    )
    print(f"Slow:\t{slow_runtime} s ({slow_runtime/num_runs} s/call)")
    fast_runtime = min(
        timeit.repeat(
            "blanc(clusters, clusters,)",
            globals={"blanc": scores.blanc, "clusters": clusters},
            number=num_runs,
            repeat=repeat,
        )
    )
    print(f"Fast:\t{fast_runtime} s ({fast_runtime/num_runs} s/call)")
    print(f"Speedup: ×{slow_runtime/fast_runtime}")


with open(tests_dir / "fixtures" / "clusters.json") as in_stream:
    clusters = scorch.clusters_from_json(in_stream)

test_metrics(clusters)

test_blanc_speedup(clusters)


tests/test_main.py
'''Unit tests for `scorch.py`.'''
import pathlib
import unittest

import pytest

from scorch import main  # noqa


@pytest.fixture()
def links():
    return [
        (1, 2),
        (2, 3),
        (4, 5),
        (5, 6),
        (4, 5),
        (5, 4),
        (7, 8),
    ]


@pytest.fixture()
def gold_file(request):
    test_file = pathlib.Path(request.module.__file__)
    return (
        test_file.resolve().parent / 'fixtures' / 'json' / 'single' / 'gold.json'
    ).open(encoding="utf8")


@pytest.fixture()
def sys_file(request):
    test_file = pathlib.Path(request.module.__file__)
    return (
        test_file.resolve().parent / 'fixtures' / 'json' / 'single' / 'sys.json'
    ).open(encoding="utf8")


@pytest.fixture()
def out_file(request):
    test_file = pathlib.Path(request.module.__file__)
    return (
        (test_file.resolve().parent / 'fixtures' / 'json' / 'single' / 'out.txt')
        .open(encoding="utf8")
        .read()
    )


@pytest.fixture()
def gold_dir(request):
    test_file = pathlib.Path(request.module.__file__)
    return test_file.resolve().parent / 'fixtures' / 'json' / 'multiple' / 'gold'


@pytest.fixture()
def sys_dir(request):
    test_file = pathlib.Path(request.module.__file__)
    return test_file.resolve().parent / 'fixtures' / 'json' / 'multiple' / 'sys'


# @pytest.fixture()
# def out_file_multiple(request):
#     test_file = pathlib.Path(request.module.__file__)
#     return (test_file.resolve().parent/'fixtures'/'json'/'multiple'/'out.txt').open().read()


def test_greedy_clustering(links):
    case = unittest.TestCase()
    expected_clusters = [{1, 2, 3}, {4, 5, 6}, {7, 8}]
    clusters = main.greedy_clustering(links)
    case.assertCountEqual(clusters, expected_clusters)


def test_process_files(gold_file, sys_file, out_file):
    expected_output = out_file
    output = ''.join(main.process_files(gold_file, sys_file))
    assert output == expected_output


# def test_process_dirs(gold_dir, sys_dir, out_file_multiple):
#     expected_output = out_file_multiple
#     output = ''.join(main.process_dirs(gold_dir, sys_dir))
#     assert output == expected_output


scorer.py
#! /usr/bin/env python3
r"""Compute CoNLL scores for coreference clustering as recommended by
*Scoring Coreference Partitions of Predicted Mentions: A Reference
Implementation* (Pradhan et al., 2014)

Usage:
  score <sys-file> <gold-file> [<out-file>]

Arguments:
  <sys-file>   system input file (json), `-` for standard input
  <gold-file>  gold input file (json), `-` for standard input
  <out-file>   output file (text), `-` for standard output [default: -]

Options:
  -h, --help  Show this screen.

Formats
The input files should be JSON files with a "type" key at top-level
  - If "type" if "graph", then top-level should have at top-level
    - A "mentions" key containing a list of singleton mention identifiers
    - A "links" key containing a list of pairs of corefering mention
      identifiers
  - If "type" is "clusters", then top-level should have a "clusters" key
    containing a mapping from clusters ids to cluster contents (as lists of
    mention identifiers).

Of course the system and gold files should use the same set of mention identifiers…

Example:
  `score sys.json gold.json out`
"""

__version__ = 'score 0.0.0'

import contextlib
import json
import signal
import sys

import itertools as it
import typing as ty

import numpy as np

from docopt import docopt
from scipy.optimize import linear_sum_assignment

# Deal with piping output in a standard-compliant way
signal.signal(signal.SIGPIPE, signal.SIG_DFL)


# Thanks http://stackoverflow.com/a/17603000/760767
@contextlib.contextmanager
def smart_open(filename: str = None, mode: str = 'r', *args, **kwargs):
    '''Open files and i/o streams transparently.'''
    if filename == '-':
        if 'r' in mode:
            stream = sys.stdin
        else:
            stream = sys.stdout
        if 'b' in mode:
            fh = stream.buffer
        else:
            fh = stream
    else:
        fh = open(filename, mode, *args, **kwargs)

    try:
        yield fh
    finally:
        try:
            fh.close()
        except AttributeError:
            pass


def greedy_clustering(links: ty.Iterable[ty.Tuple[ty.Hashable, ty.Hashable]]) -> ty.List[ty.Set]:
    '''Create transitive closure clusters from a set of edges.'''
    triaged = dict()  # type: ty.Dict[ty.Hashable, int]
    next_n = 0
    for foot, head in links:
        f_cluster = triaged.get(foot, None)
        h_cluster = triaged.get(head, None)
        if f_cluster is None:
            if h_cluster is None:
                next_n += 1
                triaged[head] = next_n
                triaged[foot] = next_n
            else:
                triaged[foot] = h_cluster
        else:
            if h_cluster is None:
                triaged[head] = f_cluster
            else:
                if f_cluster != h_cluster:  # merge `f_cluster` and `h_cluster`
                    for e, c in triaged.items():
                        if c == f_cluster:
                            triaged[e] = h_cluster
    clusters = dict()  # type: ty.Dict[int, ty.List]
    for e, n in triaged.items():
        c = clusters.get(n, None)
        if c is None:
            clusters[n] = [e]
        else:
            c.append(e)
    return [set(c) for c in clusters.values()]


def clusters_from_graph(nodes: ty.Iterable[ty.Hashable],
                        edges: ty.Iterable[ty.Tuple[ty.Hashable, ty.Hashable]]) -> ty.List[ty.Set]:
    clusters = greedy_clustering(edges)
    non_sing = set.union(*clusters)
    singletons = [{n} for n in nodes if n not in non_sing]
    clusters.extend(singletons)
    return clusters


def clusters_from_json(fp) -> ty.List[ty.Set]:
    obj = json.load(fp)
    if obj["type"] == "graph":
        return clusters_from_graph(obj['mentions'], obj['links'])
    elif obj["type"] == "clusters":
        return [set(c) for c in obj["clusters"]]
    raise ValueError('Unsupported input format')


# OPTIMIZE: This could be made faster by dealing separately with singletons
# OPTIMIZE: This could be made faster (in average) by sorting `cluster_lst` by decreasing length
def links_from_clusters(clusters: ty.Iterable[ty.Set]
                        ) -> ty.Tuple[ty.List[ty.Tuple[ty.Hashable, ty.Hashable]],
                                      ty.List[ty.Tuple[ty.Hashable, ty.Hashable]]]:
    clusters_lst = list(clusters)
    elements = sorted(set.union(*clusters_lst))
    C = []
    N = []
    for i, j in it.combinations(elements, 2):
        if i == j:
            continue
        elif any(c for c in clusters_lst if i in c and j in c):
            C.append((i, j))
        else:
            N.append((i, j))
    return C, N


def trace(cluster: ty.Set, partition: ty.Iterable[ty.Set]) -> ty.Iterable[ty.Set]:
    r'''
    Return the partition of `#cluster` induced by `#partition`, that is
    ```math
    \{C∩A|A∈P\} ∪ \{\{x\}|x∈C∖∪P\}
    ```
    Where `$C$` is `#cluster` and `$P$` is `#partition`.

    This assume that the elements of `#partition` are indeed pairwise disjoint'''
    remaining = set(cluster)
    for a in partition:
        common = remaining.intersection(a)
        if common:
            remaining.difference_update(common)
            yield common
    for x in sorted(remaining):
        yield set((x,))


def muc(key: ty.List[ty.Set], response: ty.List[ty.Set]) -> ty.Tuple[float, float, float]:
    r'''
    Compute the MUC `$(R, P, F₁)$` scores for a `#response` clustering given a `#key`
    clustering`, that is
    ```math
    R &= \frac{∑_{k∈K}(\#k-\#p(k, R))}{∑_{k∈K}(\#k-1)}\\
    P &= \frac{∑_{r∈R}(\#r-\#p(r, K))}{∑_{r∈R}(\#r-1)}\\
    F &= 2*\frac{PR}{P+R}
    ```
    with `$p(x, E)=\{x∩A|A∈E\}$`
    '''
    R = sum(len(k) - sum(1 for _ in trace(k, response)) for k in key)/sum(len(k)-1 for k in key)
    P = sum(len(r)-sum(1 for _ in trace(r, key)) for r in response)/sum(len(r)-1 for r in response)
    F = (2*P*R)/(P+R)
    return R, P, F


def b_cubed(key: ty.List[ty.Set], response: ty.List[ty.Set]) -> ty.Tuple[float, float, float]:
    '''
    Compute the B³ `$(R, P, F₁)$` scores for a `#response` clustering given a `#key`
    clustering`, that is
    ```math
    R &= \frac{∑_{k∈K}∑_{\∈R}\frac{(\#k∩r)²}{#k}}{∑_{k∈K}\#k}\\
    P &= \frac{∑_{r∈R}∑_{k∈K}\frac{(\#r∩k)²}{#r}}{∑_{r∈R}\#r}\\
    F &= 2*\frac{PR}{P+R}
    ```
    '''
    R = (sum(len(k.intersection(r))**2/len(k) for k in key for r in response) /
         sum(len(k) for k in key))
    P = (sum(len(r.intersection(k))**2/len(r) for r in response for k in key) /
         sum(len(r) for r in response))
    F = (2*P*R)/(P+R)
    return R, P, F


def ceaf(key: ty.List[ty.Set],
         response: ty.List[ty.Set],
         score: ty.Callable[[ty.Set, ty.Set], float]) -> ty.Tuple[float, float, float]:
    r'''
    Compute the CEAF `$(R, P, F₁)$` scores for a `#response` clustering given a `#key`
    clustering` using the `#score` alignment score function, that is
    ```math
    R &= \frac{∑_{k∈K}C(k, A(k))}{∑_{k∈K}\#k}\\
    P &=  \frac{∑_{r∈R}C(r, A⁻¹(r))}{∑_{r∈R}\#r}\\
    F &= 2*\frac{PR}{P+R}
    ```
    Where `$C$` is `#score` and `$A$` is the one-to-one mapping from key clusters to
    response clusters that maximizes `$∑_{k∈K}C(k, A(k))$`.
    '''
    cost_matrix = np.array([[-score(k, r) for r in response] for k in key])
    row_ind, col_ind = linear_sum_assignment(cost_matrix)
    total_score = -cost_matrix[row_ind, col_ind].sum()
    print(total_score, sum(len(k) for k in key))
    R = total_score/sum(score(k, k) for k in key)
    P = total_score/sum(score(r, r) for r in response)
    F = (2*P*R)/(P+R)
    return R, P, F


def ceaf_m(key: ty.List[ty.Set], response: ty.List[ty.Set]) -> ty.Tuple[float, float, float]:
    r'''
    Compute the CEAF_m `$(R, P, F₁)$` scores for a `#response` clustering given a `#key`
    clustering`, that is the CEAF score for the `$Φ_3$` score function
    ```math
    Φ_3: (k, r) ⟼ \#k∩r
    ```
    '''
    def Φ_3(k, r):
        return len(k.intersection(r))
    return ceaf(key, response, Φ_3)


def ceaf_e(key: ty.List[ty.Set], response: ty.List[ty.Set]) -> ty.Tuple[float, float, float]:
    r'''
    Compute the CEAF_m `$(R, P, F₁)$` scores for a `#response` clustering given a `#key`
    clustering`, that is the CEAF score for the `$Φ₄$` score function
    (aka the Sørensen–Dice coefficient).
    ```math
    Φ₄: (k, r) ⟼ \frac{2×\#k∩r}{\#k+\#r}
    Note: this use the original (Luo, 2005) definition as opposed to the (Pradhan et al. 2014) one
    which inlines the denominators.
    ```
    '''
    def Φ_4(k, r):
        return 2*len(k.intersection(r))/(len(k)+len(r))
    return ceaf(key, response, Φ_4)


# COMBAK: Check the numeric stability
def blanc(key: ty.List[ty.Set], response: ty.List[ty.Set]) -> ty.Tuple[float, float, float]:
    r'''
    Return the BLANC `$(R, P, F)$` scores for a `#response` clustering given a `#key`
    clustering`, that is.
    '''
    C_k, N_k = map(set, links_from_clusters(key))
    C_r, N_r = map(set, links_from_clusters(response))

    T_c = C_k.intersection(C_r)
    T_n = N_k.intersection(N_r)

    R_c, P_c = len(T_c)/len(C_k), len(T_c)/len(C_r)
    half_F_c = P_c*R_c/(P_c+R_c)  # No need to multiply this by 2 to divide right after

    R_n, P_n = len(T_n)/len(N_k), len(T_n)/len(N_r)
    half_F_n = P_n*R_n/(P_n+R_n)

    return (R_c+R_n)/2, (P_c+P_n)/2, half_F_c+half_F_n


# def main_entry_point(argv=None):
#     arguments = docopt(__doc__, version=__version__, argv=argv)
#     # Since there are no support for default positional arguments in
#     # docopt yet. Might be useful for complex default values, too
#     if arguments['<out-file>'] is None:
#         arguments['<out-file>'] = '-'
#
#     with smart_open(arguments['<sys-input>']) as in_stream:
#         sys_clusters = clusters_from_json(in_stream)
#
#     with smart_open(arguments['<gold-input>']) as in_stream:
#         gold_clusters = clusters_from_json(in_stream)
#
#     R, P, F = muc(sys_clusters, gold_clusters)
#
#     with smart_open(arguments['<out-file>'], 'w') as out_stream:
#         out_stream.write(f'MUC Score\nP: {P}\tR: {R}\tF₁: {F}\n')
#
#
# if __name__ == '__main__':
#     sys.exit(main_entry_point())


scorch/main.py
#! /usr/bin/env python3
r"""Compute CoNLL scores for coreference clustering as recommended by
*Scoring Coreference Partitions of Predicted Mentions: A Reference
Implementation* (Pradhan et al., 2014)

## Usage:
  scorch <gold> <sys> [<out-file>]

## Arguments:
  <gold>      gold input file (json) or directory, `-` for standard input
  <sys>       system input file (json) or directory, `-` for standard input
  <out-file>  output file (text), `-` for standard output [default: -]

## Options:
  -h, --help       Show this screen.

## Example:
  `scorch gold.json sys.json out.txt`
  `scorch gold/ sys/ out.txt`
"""

import contextlib
import json
import pathlib
import sys

import typing as ty

from statistics import mean

import tqdm

import numpy as np

from docopt import docopt


from scorch import scores
from scorch import __version__


METRICS = {
    "MUC": scores.muc,
    "B³": scores.b_cubed,
    "CEAF_m": scores.ceaf_m,
    "CEAF_e": scores.ceaf_e,
    "BLANC": scores.blanc,
}


# Thanks http://stackoverflow.com/a/17603000/760767
@contextlib.contextmanager
def smart_open(
    filename: ty.Union[str, pathlib.Path], mode: str = "r", *args, **kwargs
) -> ty.Generator[ty.IO, None, None]:
    """Open files and i/o streams transparently."""
    fh: ty.IO
    if filename == "-":
        if "r" in mode:
            stream = sys.stdin
        else:
            stream = sys.stdout
        if "b" in mode:
            fh = stream.buffer
        else:
            fh = stream
    else:
        fh = open(filename, mode, *args, **kwargs)

    try:
        yield fh
    finally:
        try:
            fh.close()
        except AttributeError:
            pass


def greedy_clustering(
    links: ty.Iterable[ty.Tuple[ty.Hashable, ty.Hashable]]
) -> ty.List[ty.Set]:
    r"""
    Get connected component clusters from a set of edges.

    This is basically done with depth 1 [disjoint-sets][1], which seemed a good trade-off between
    performance and readability. If performance is an issue, consider using e.g. the [incremental
    connected component algorithm][2] found in Boost Graph.

    Even better: don't rely on this and preprocess your data using a more clever clustering
    procedure, e.g. with [NetworkX][3].

    [1]: https://en.wikipedia.org/wiki/Disjoint-set_data_structure
    [2]: http://www.boost.org/doc/libs/1_66_0/libs/graph/doc/incremental_components.html
    [3]: https://networkx.github.io/documentation/networkx-1.9.1/reference/algorithms.mst.html
    """
    # The idea is to build an element→cluster mapping. This way, when we get a a→b link, we just
    # add a to the cluster of b.
    clusters: ty.Dict[ty.Hashable, ty.List] = dict()
    # The issue with this is that it is tedious to deduplicate afterward since Python `set` can't
    # be passed a custom key.
    # So instead we use the first element we encounter in each cluster as its head, and keep two
    # mappings: element→head and head→cluster.
    heads: ty.Dict[ty.Hashable, ty.Hashable] = dict()
    for source, target in links:
        source_head = heads.setdefault(source, source)
        source_cluster = clusters.setdefault(source_head, [source_head])

        target_head = heads.get(target)
        if target_head is None:
            heads[target] = source_head
            source_cluster.append(target)
        elif (
            target_head is not source_head
        ):  # Merge `target`'s cluster into `source`'s'
            for e in clusters[target_head]:
                heads[e] = source_head
            source_cluster.extend(clusters[target_head])
            del clusters[target_head]

    return [set(c) for c in clusters.values()]


def clusters_from_graph(
    nodes: ty.Iterable[ty.Hashable],
    edges: ty.Iterable[ty.Tuple[ty.Hashable, ty.Hashable]],
) -> ty.List[ty.Set]:
    """Return the connex components of a graph."""
    clusters = greedy_clustering(edges)
    if not clusters:
        return [{n} for n in nodes]
    non_sing = set.union(*clusters)
    singletons = [{n} for n in nodes if n not in non_sing]
    clusters.extend(singletons)
    return clusters


def clusters_from_json(fp) -> ty.List[ty.Set]:
    obj = json.load(fp)
    if obj["type"] == "graph":
        return clusters_from_graph(obj["mentions"], obj["links"])
    elif obj["type"] == "clusters":
        return [set(c) for n, c in obj["clusters"].items()]
    raise ValueError("Unsupported input format")


def process_files(
    gold_fp: ty.TextIO, sys_fp: ty.TextIO, add_sys_mentions=True
) -> ty.Iterable[str]:
    gold_clusters = clusters_from_json(gold_fp)
    sys_clusters = clusters_from_json(sys_fp)
    if add_sys_mentions:
        gold_mentions: ty.Set[ty.Hashable] = set(m for c in gold_clusters for m in c)
        sys_mentions: ty.Set[ty.Hashable] = set(m for c in sys_clusters for m in c)
        extra_mentions = sys_mentions - gold_mentions
        gold_clusters.extend(set([m]) for m in extra_mentions)
    for name, metric in METRICS.items():
        R, P, F = metric(gold_clusters, sys_clusters)
        yield f"{name}:\tR={R}\tP={P}\tF₁={F}\n"
    conll_score = scores.conll2012(gold_clusters, sys_clusters)
    yield f"CoNLL-2012 average score: {conll_score}\n"


def process_dirs(gold_dir, sys_dir, add_sys_mentions=True) -> ty.Iterable[str]:
    gold_path = pathlib.Path(gold_dir)
    sys_path = pathlib.Path(sys_dir)
    pairs = dict()  # ty.Dict[str, ty.Tuple[pathlib.Path, pathlib.Path]]
    for sys_file in sys_path.iterdir():
        try:
            gold_file = next(gold_path.glob(f"{sys_file.stem}*"))
        except StopIteration:
            raise ValueError(f"No matching gold file for {sys_file}")
        pairs[sys_file.stem] = (gold_file, sys_file)

    individual_results = (
        []
    )  # ty.List[str, ty.Dict[str, ty.Tuple[float, float, float]] int, int]
    pbar = tqdm.tqdm(
        pairs.items(),
        unit="document",
        desc="Scoring",
        unit_scale=True,
        unit_divisor=1024,
        dynamic_ncols=True,
        leave=False,
    )
    for name, (gold_file, sys_file) in pbar:
        pbar.desc = f"Scoring {name}"
        with gold_file.open() as gold_stream, sys_file.open() as sys_stream:
            gold_clusters = clusters_from_json(gold_stream)
            sys_clusters = clusters_from_json(sys_stream)
            if add_sys_mentions:
                gold_mentions: ty.Set[ty.Hashable] = set(
                    m for c in gold_clusters for m in c
                )
                sys_mentions: ty.Set[ty.Hashable] = set(
                    m for c in sys_clusters for m in c
                )
                extra_mentions = sys_mentions - gold_mentions
                gold_clusters.extend(set([m]) for m in extra_mentions)
        r = {
            name: metric(gold_clusters, sys_clusters)
            for name, metric in METRICS.items()
        }
        individual_results.append(
            (name, r, sum(map(len, gold_clusters)), sum(map(len, sys_clusters)))
        )

    gold_sizes = np.fromiter(
        (g_size for *_, g_size, _ in individual_results), dtype=int
    )
    sys_sizes = np.fromiter((s_size for *_, _, s_size in individual_results), dtype=int)

    results = {}
    for name in METRICS:
        all_R, all_P, all_F = (
            np.fromiter(s, float)
            for s in zip(*(r[name] for _, r, *_ in individual_results))
        )
        R = np.average(all_R, weights=gold_sizes)
        P = np.average(all_P, weights=sys_sizes)
        F = np.average(all_F, weights=gold_sizes + sys_sizes)
        results[name] = (R, P, F)
        yield f"{name}:\tR={R}\tP={P}\tF₁={F}\n"
    conll_score = mean(results[s][2] for s in ("MUC", "B³", "CEAF_e"))
    yield f"CoNLL-2012 average score: {conll_score}\n"


def main_entry_point(argv=None):
    arguments = docopt(__doc__, version=__version__, argv=argv)
    # Since there are no support for default positional arguments in
    # docopt yet. Might be useful for complex default values, too
    if arguments["<out-file>"] is None:
        arguments["<out-file>"] = "-"

    if arguments["<gold>"] != "-" and arguments["<sys>"] != "-":
        gold_path = pathlib.Path(arguments["<gold>"])
        sys_path = pathlib.Path(arguments["<sys>"])
        if gold_path.is_dir() and sys_path.is_dir():
            with smart_open(arguments["<out-file>"], "w") as out_stream:
                out_stream.writelines(process_dirs(gold_path, sys_path))
            return None

    with contextlib.ExitStack() as stack:
        gold_stream = stack.enter_context(smart_open(arguments["<gold>"]))
        sys_stream = stack.enter_context(smart_open(arguments["<sys>"]))
        out_stream = stack.enter_context(smart_open(arguments["<out-file>"], "w"))
        out_stream.writelines(process_files(gold_stream, sys_stream))


if __name__ == "__main__":
    sys.exit(main_entry_point())
